### PR TITLE
use exponentional backoff and jitter when polling tasks

### LINKED ISSUE #17: Proper retry, timeout, etc..
Regarding retries and timeouts: We should have both sane default behavior and everything controllable if user wants to fine-tune.

We need:
- retry with exponential backoff, maybe capped, with jitter
- timeout: both on each network call, and globally on the retry: we don't want to wait indefinitely because of infinite retry.

### PR DESCRIPTION
**Pull request by maingoh**
Close #17

### CODE REVIEW

File: deepomatic/resources/task.py [hunk 1]
```diff
@@ -40,30 +49,28 @@ def list(self, task_ids):
...
-            sleep_time = time.time() - (last_time + sleep_time)
-            if sleep_time > 0:
-                time.sleep(sleep_time)
-            self.refresh()
+        return self
 
-        if self['status'] == "error":
-            raise TaskError(self.data())
+    def _refresh_status(self):
+        self.refresh()
```
Comment (maingoh) [inline deepomatic/resources/task.py]: I am wondering if we should we refresh is the status is not pending ?

File: deepomatic/resources/task.py [hunk 2]
```diff
@@ -40,30 +49,28 @@ def list(self, task_ids):
...
-        last_time = start_time
-        sleep_time = 0.2
-        while self['status'] == "pending":
-            # Check for timeout
-            if time.time() > start_time + timeout:
+        try:
+            retryer = Retrying(wait=wait_random_exponential(multiplier=wait_exp_multiplier, max=wait_exp_max),
+                               stop=stop_after_delay(timeout),
+                               retry=retry_if_result(is_pending_status),
+                               before=before_log(logger, logging.DEBUG),
```
Comment (thomas-riccardi) [inline deepomatic/resources/task.py]: maybe create a sublogger for `task.wait` ? with a context logging the task id?

File: deepomatic/resources/task.py [hunk 3]
```diff
@@ -40,30 +50,29 @@ def list(self, task_ids):
...
+                               after=after_log(logger, logging.DEBUG))
+            retryer(self._refresh_status)
+        except RetryError:
+            if self['status'] == 'pending':
                 raise TaskTimeout(self.data())
+            elif self['status'] == 'error':
```
Comment (vdel) [inline deepomatic/resources/task.py]: What happens if the status is 'error' ? Is RetryError thrown ?

**Review by thomas-riccardi — APPROVED** [review]

File: deepomatic/resources/task.py (hunk 3 above)
Comment (maingoh) [reply deepomatic/resources/task.py]: Indeed, I made a mistake, the retry is done only if status is pending, I just fixed it.

**Pull request merged**